### PR TITLE
API-Refactor: Wire up bundle and entry caches into new API

### PR DESCRIPTION
**Pull request by marcosy**
**Pull Request check list**

- [x] Commit conforms to CONTRIBUTING.md?
- [x] Proper tests/regressions included?
- [x] Documentation updated?

**Affected functionality**

New API cache

**Description of change**

This PR wires up the bundle and registration entries caches into the new API. The caches used by the old API were moved to a common package in order to reuse them.

~This is still WIP (unit test are missing), but I’m pushing this PR as draft in order to get some early feedback and validate the approach.~

**Which issue this PR fixes**

Fixes: #1631

Signed-off-by: Marcos Yedro <marcosyedro@gmail.com>

### CODE REVIEW

**Comment by marcosy** [comment]
I've kept the `disable_cache` flag in addition to the cache invalidation for `AppendBundle` and `PublishJWTAuthority`. It looks to me that it may be convenient to have an option to force a fresh fetch instead of retrieving one from cache (default).

However, I am happy to remove it if we think that it is not any use case for this.

File: pkg/server/endpoints/config.go [hunk 1]
```diff
@@ -144,13 +148,13 @@ func (c *Config) maybeMakeExperimentalServers() *ExperimentalServers {
 		EntryServer: entryv1.New(entryv1.Config{
 			TrustDomain:  c.TrustDomain,
 			DataStore:    ds,
-			EntryFetcher: authorizedEntryFetcher,
+			EntryFetcher: authorizedEntryFetcherWithCache,
```
[the remark this answers is not in the document]
Comment (marcosy) [reply pkg/server/endpoints/config.go:151]: Good question. The old API used the entries cache for the `FetchX509SVID()` RPC but it did not use it for the`FetchJWTSVID()` RPC.

In the new API, the entries cache is part of the service so it is like an "all or none" situation. At first, I did not include it. But thinking about it again, I don't see a reason for not enabling the cache in the SVID service too.

The cache expires after 1 second, so I believe it is an acceptable time to wait after updating a registration entry to be reflected in an SVID rotation.

Thoughts @azdagron @amartinezfayo ?

**Comment by azdagron** [comment]
I think we're a large contributor of bundle changes, bundle appending done by the CA manager. Other layers could also benefit from the bundle cache, like the TLS configuration done by the endpoints package, which fetches the bundle fresh on every TLS handshake.

It may be worth stepping back and looking at the datastore FetchBundle usage across the codebase. We might be able to implement this all by putting a shim around the datastore at a higher layer which caches FetchBundle responses for up to a second, if other bundle modifying methods were called, it could immediately invalidate the cache. We could make it opt-in, perhaps by supplying a special value stuffed into the context so that we were safe and consistent by default (not all FetchBundle call sites can tolerate eventual consistency, e.g. the IdentityProvider hostservice used by bundle notifiers).

**Comment by marcosy** [comment]
That sounds good to me! 
 
I wrapped the datastore implementation adding a cache layer. I believe this approach is simple and allows us to cache other entities in the future in an easy way.
 
 Since we are close to a release, I was a bit conservative on the cache introduction. I kept the `FetchBundle` calls that were previously cached and introduced only a couple of new ones (search for `.WithCache`)
 
 Open to comments and suggestions!

File: proto/spire/api/server/bundle/v1/bundle.proto [hunk 2]
```diff
@@ -63,6 +63,9 @@ service Bundle {
 message GetBundleRequest {
     // An output mask indicating which bundle fields are set in the response.
     spire.types.BundleMask output_mask = 1;
+
+    // Bool value to determinate weather to use cached results or not
+    bool disable_cache = 2;
```
Comment (azdagron) [inline proto/spire/api/server/bundle/v1/bundle.proto:68]: I think we were going to back this out, right?

File: pkg/server/endpoints/config.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/server/endpoints/config.go:150]: I think we should use it for both.

File: go.mod [hunk 3]
```diff
@@ -56,6 +56,7 @@ require (
 	github.com/shirou/gopsutil v2.18.12+incompatible
 	github.com/shirou/w32 v0.0.0-20160930032740-bb4de0191aa4 // indirect
 	github.com/sirupsen/logrus v1.4.2
+	github.com/spiffe/go-spiffe v1.1.0
```
[the remark this answers is not in the document]
Comment (marcosy) [reply go.mod:59]: uups! right it is only because of the spiffetest pkg, I'll remove it

File: proto/spire/api/server/bundle/v1/bundle.proto (hunk 2 above)
Comment (marcosy) [reply proto/spire/api/server/bundle/v1/bundle.proto:68]: Right! I forgot to remove it from the previous approach.

File: pkg/server/cache/dscache/cache_test.go [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+}
+
+func TestBundleInvalidations(t *testing.T) {
+	req := &datastore.FetchBundleRequest{TrustDomainId: "spiffe://domain.test"}
+	bundle1, bundle2 := getBundles(t, "spiffe://domain.test")
+
+	for _, tt := range []struct {
+		name             string
+		invalidatingFunc func(cache *DatastoreCache)
+		dsFailiure       bool
```
Comment (azdagron) [inline pkg/server/cache/dscache/cache_test.go:88]:
```suggestion
		dsFailure       bool
```

**Pull request merged**